backend/jobs/codeAnalysisJob.ts

```diff
@@ -0,0 +1,370 @@
+
+  const courses = await CourseModel.find({
+    installationId: { $in: installationIds },
+  });
+
+  await Promise.all(
+    courses.map(async course => {
+      if (course && course.installationId) {
+        const installationOctokit = await app.getInstallationOctokit(
+          course.installationId
+        );
+        await getCourseCodeData(installationOctokit, course);
+      }
+    })
+  );
+
+  console.log('Code analysis data fetched');
+};
+
+const getCourseCodeData = async (octokit: Octokit, course: any) => {
+  if (!course.gitHubOrgName) return;
+  const gitHubOrgName = course.gitHubOrgName;
+
+  const repos = await octokit.rest.repos.listForOrg({
+    org: course.gitHubOrgName,
+    sort: 'updated',
+    per_page: 300,
+    direction: 'desc',
+  });
+  let allRepos = repos.data;
+
+  if (course.repoNameFilter) {
+    allRepos = allRepos.filter(repo =>
+      repo.name.includes(course.repoNameFilter as string)
+    );
+  }
+
+  for (const repo of allRepos) {
+    const buildTool = await getRepoBuildTool(octokit, gitHubOrgName, repo.name);
+    if (buildTool === '.NET' || buildTool === 'Error') continue;
+
+    await createProjectIfNotExists(gitHubOrgName + '_' + repo.name);
+
+    await getLatestCommit(gitHubOrgName, repo.name);
+
+    await scanRepo(gitHubOrgName, repo.name, buildTool);
+
+    await getAndSaveCodeData(gitHubOrgName, repo);
+  }
+};
+
+const getRepoBuildTool = async (
+  octokit: Octokit,
+  owner: string,
+  repo: string
+) => {
+  try {
+    // Fetch repository root directory content
+    const contents = await octokit.rest.repos.getContent({
+      owner,
+      repo,
+      path: '',
+    });
+
+    if (Array.isArray(contents.data)) {
+      const files = contents.data.map(file => file.name);
+
+      // Check for Maven
+      if (files.includes('pom.xml')) {
+        return 'Maven';
+      }
+
+      // Check for Gradle
+      if (
+        files.includes('build.gradle') ||
+        files.includes('build.gradle.kts')
+      ) {
+        return 'Gradle';
+      }
+
+      // Check for .NET-related project files
+      const dotnetProjectFiles = files.filter(
+        file =>
+          file.endsWith('.csproj') ||
+          file.endsWith('.fsproj') ||
+          file.endsWith('.vbproj') ||
+          file.endsWith('.sln')
+      );
+
+      if (dotnetProjectFiles.length > 0) return '.NET';
+
+      // If none of the above, return 'Others'
+      return 'Others';
+    }
+
+    return 'Others';
+  } catch (error) {
+    if (error instanceof Error) {
+      console.error(
+        `Error fetching build tool for repository ${repo}: ${error.message}`
+      );
+    } else {
+      console.error(
+        `Unknown error fetching build tool for repository ${repo}:`,
+        error
+      );
+    }
+    return 'Error';
+  }
+};
+
+const createProjectIfNotExists = async (repo: string) => {
+  try {
+    const sonarUri = process.env.SONAR_URI;
```

Comment:
Is `sonarUri` a fixed string? If it is just a fixed string value, we can put it in `endpoints.ts` instead of as an environment variable.

Reply:
During dev, I first test it out on localhost, before testing it out on staging so I do change the sonarUri. On staging and eventually prod itll be localhost:9000